### ISSUE TITLE
I don't know why the LoRA for training failed

### ISSUE THREAD

**Issue by chenminglin** [kijai/ComfyUI-FluxTrainer#32]
The training validation image is black, and the loss map has no data。
![workflow (2)](https://github.com/user-attachments/assets/c87a7841-c21a-4b75-8327-5f11aa9f8f07)
2024-09-03 00:16:48 INFO     got prompt                           server.py:522
2024-09-03 00:16:48 INFO     t5xxl_max_token_le flux_train_network_comfy.py:113
                             ngth: 512
You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
2024-09-03 00:16:50 INFO     Loading dataset config from   train_network.py:273
                             [[datasets]]
                             resolution = [ 512, 512,]
                             batch_size = 1
                             enable_bucket = true
                             bucket_no_upscale = false
                             min_bucket_reso = 256
                             max_bucket_reso = 1024
                             [[datasets.subsets]]
                             image_dir =
                             "E:\\train_mater\\hwan_02\\01
                             _hwan_ollama"
                             class_tokens = "hwan"
                             num_repeats = 10

                             [general]
                             shuffle_caption = false
                             caption_extension = ".txt"
                             keep_tokens_separator = "|||"
                             caption_dropout_rate = 0.0
                             color_aug = false
                             flip_aug = false

                    INFO     prepare images.                 train_util.py:1814
                    INFO     get image size from name of     train_util.py:1748
                             cache files
                    INFO     set image size from cache       train_util.py:1755
                             files: 14/14
                    INFO     found directory                 train_util.py:1757
                             E:\train_mater\hwan_02\01_hwan_
                             ollama contains 14 image files
                    INFO     140 train images with           train_util.py:1855
                             repeating.
                    INFO     0 reg images.                   train_util.py:1858
                    WARNING  no regularization images /      train_util.py:1863
                             正則化画像が見つかりませんでし
                             た
                    INFO     [Dataset 0]                     config_util.py:570
                               batch_size: 1
                               resolution: (512, 512)
                               enable_bucket: True
                               network_multiplier: 1.0
                               min_bucket_reso: 256
                               max_bucket_reso: 1024
                               bucket_reso_steps: 64
                               bucket_no_upscale: False

                               [Subset 0 of Dataset 0]
                                 image_dir:
                             "E:\train_mater\hwan_02\01_hwan
                             _ollama"
                                 image_count: 14
                                 num_repeats: 10
                                 shuffle_caption: False
                                 keep_tokens: 0
                                 keep_tokens_separator: |||
                                 caption_separator: ,
                                 secondary_separator: None
                                 enable_wildcard: False
                                 caption_dropout_rate: 0.0
                                 caption_dropout_every_n_epo
                             ches: 0
                                 caption_tag_dropout_rate:
                             0.0
                                 caption_prefix: None
                                 caption_suffix: None
                                 color_aug: False
                                 flip_aug: False
                                 face_crop_aug_range: None
                                 random_crop: False
                                 token_warmup_min: 1,
                                 token_warmup_step: 0.0,
                                 alpha_mask: False,
                                 is_reg: False
                                 class_tokens: hwan
                                 caption_extension: .txt

                    INFO     [Dataset 0]                     config_util.py:576
                    INFO     loading image sizes.             train_util.py:876
                    INFO     make buckets                     train_util.py:882
                    INFO     number of images (including      train_util.py:928
                             repeats) /
                             各bucketの画像枚数（繰り返
                             し回数を含む）
                    INFO     bucket 0: resolution (512, 512), train_util.py:933
                             count: 140
                    INFO     mean ar error (without repeats): train_util.py:938
                             0.0
                    INFO     preparing accelerator         train_network.py:338
accelerator device: cuda
                    INFO     Building Flux model dev           flux_utils.py:22
                    INFO     Loading state dict from           flux_utils.py:27
                             E:\ComfyUI-aki-v1.3\models\unet\f
                             lux\flux1-dev-fp8.safetensors
                    INFO     Loaded Flux: <All keys matched    flux_utils.py:30
                             successfully>
                    INFO     Building CLIP                     flux_utils.py:47
                    INFO     Loading state dict from          flux_utils.py:140
                             E:\ComfyUI-aki-v1.3\models\clip\
                             sd3\clip_l.safetensors
                    INFO     Loaded CLIP: <All keys matched   flux_utils.py:143
                             successfully>
                    INFO     Loading state dict from          flux_utils.py:186
                             E:\ComfyUI-aki-v1.3\models\clip\
                             sd3\t5xxl_fp8_e4m3fn.safetensors
                    INFO     Loaded T5xxl: <All keys matched  flux_utils.py:189
                             successfully>
                    INFO     Building AutoEncoder              flux_utils.py:35
                    INFO     Loading state dict from           flux_utils.py:39
                             E:\ComfyUI-aki-v1.3\models\vae\fl
                             ux_ae.safetensors
                    INFO     Loaded AE: <All keys matched      flux_utils.py:42
                             successfully>
import network module: .networks.lora_flux
2024-09-03 00:16:51 INFO     [Dataset 0]                     train_util.py:2337
                    INFO     caching latents with caching     train_util.py:984
                             strategy.
                    INFO     checking cache validity...       train_util.py:994
                    INFO     no latents to cache             train_util.py:1034
                    INFO     move vae and unet  flux_train_network_comfy.py:156
                             to cpu to save
                             memory
                    INFO     move text encoders flux_train_network_comfy.py:164
                             to gpu
2024-09-03 00:16:58 INFO     [Dataset 0]                     train_util.py:2358
                    INFO     caching Text Encoder outputs    train_util.py:1111
                             with caching strategy.
                    INFO     checking cache validity...      train_util.py:1117
                    INFO     no Text Encoder outputs to      train_util.py:1139
                             cache
                    INFO     cache Text Encoder flux_train_network_comfy.py:173
                             outputs for sample
                             prompt: ['a
                             woman,hwan']
                    INFO     cache Text Encoder flux_train_network_comfy.py:203
                             outputs for
                             prompt: a
                             woman,hwan
                    INFO     cache Text Encoder flux_train_network_comfy.py:203
                             outputs for
                             prompt:
                    INFO     move CLIP-L back   flux_train_network_comfy.py:212
                             to cpu
2024-09-03 00:16:59 INFO     move t5XXL back to flux_train_network_comfy.py:214
                             cpu
2024-09-03 00:17:07 INFO     move vae and unet  flux_train_network_comfy.py:219
                             back to original
                             device
                    INFO     create LoRA network. base dim     lora_flux.py:367
                             (rank): 16, alpha: 1.0
                    INFO     neuron dropout: p=None, rank      lora_flux.py:368
                             dropout: p=None, module dropout:
                             p=None
                    INFO     create LoRA for Text Encoder 1:   lora_flux.py:441
                    INFO     create LoRA for Text Encoder 2:   lora_flux.py:441
                    INFO     create LoRA for Text Encoder: 72  lora_flux.py:446
                             modules.
2024-09-03 00:17:08 INFO     create LoRA for FLUX all blocks:  lora_flux.py:458
                             304 modules.
                    INFO     enable LoRA for U-Net: 304        lora_flux.py:501
                             modules
FLUX: Gradient checkpointing enabled. CPU offload: False
prepare optimizer, data loader etc.
                    INFO     use 8-bit AdamW optimizer | {}  train_util.py:4353
enable full fp16 training.
enable fp8 training for U-Net.
running training
  num train images * repeats: 140
  num reg images: 0
  num batches per epoch: 140
  num epochs: 11
  batch size per device: 1
  gradient accumulation steps: 1
  total optimization steps: 1500
2024-09-03 00:17:20 INFO     text_encoder is not needed   train_network.py:1017
                             for training. deleting to
                             save memory.
                    INFO     unet dtype:                  train_network.py:1034
                             torch.float8_e4m3fn, device:
                             cuda:0

**Comment by kijai** [comment]
Is there a reason you chose fp16 as gradient dtype? I can't really think of anything else, what kind of hardware are you using?

**Comment by chenminglin** [comment]
After I tried to modify the gradient dtype, it finally worked, thank you very much. My graphics card is 4070Ti super.
![WX20240903-123202@2x](https://github.com/user-attachments/assets/5fda2218-c5ff-40a1-b09b-e206dbd19b66)

**Comment by muxin159** [comment]
> Is there a reason you chose fp16 as gradient dtype? I can't really think of anything else, what kind of hardware are you using?您选择 fp16 作为梯度数据类型有什么原因吗？我实在想不出其他的了，你用的是什么类型的硬件？
Hello KJ,First of all, thank you for your outstanding work, which has provided me with a highly efficient interface for training LoRA. However I have encountered an issue. When I select fp16, I experience the same situation as the author of this problem. But when I select bf16, ComfyUI prompts me with the message: “bf16 mixed precision requires PyTorch >= 1.10 and a supported device.” I am using PyTorch 2.1.2, and my graphics card is 2080Ti with 22GB, which has left me quite confused. If you have time, I would appreciate your response. Once again, thank you for your excellent work.

**Comment by kijai** [comment]
> > Is there a reason you chose fp16 as gradient dtype? I can't really think of anything else, what kind of hardware are you using?您选择 fp16 作为梯度数据类型有什么原因吗？我实在想不出其他的了，你用的是什么类型的硬件？
> > Hello KJ,First of all, thank you for your outstanding work, which has provided me with a highly efficient interface for training LoRA. However I have encountered an issue. When I select fp16, I experience the same situation as the author of this problem. But when I select bf16, ComfyUI prompts me with the message: “bf16 mixed precision requires PyTorch >= 1.10 and a supported device.” I am using PyTorch 2.1.2, and my graphics card is 2080Ti with 22GB, which has left me quite confused. If you have time, I would appreciate your response. Once again, thank you for your excellent work.

Pytorch 2.1.2 is also too old and kohya himself recommends 2.4.0 minimum, I'd recommend updating to 2.4.1. Versions before that also use a lot more VRAM so it's pretty much necessary.

**Comment by muxin159** [comment]
> > > Is there a reason you chose fp16 as gradient dtype? I can't really think of anything else, what kind of hardware are you using?您选择 fp16 作为梯度数据类型有什么原因吗？我实在想不出其他的了，你用的是什么类型的硬件？
> > > Hello KJ,First of all, thank you for your outstanding work, which has provided me with a highly efficient interface for training LoRA. However I have encountered an issue. When I select fp16, I experience the same situation as the author of this problem. But when I select bf16, ComfyUI prompts me with the message: “bf16 mixed precision requires PyTorch >= 1.10 and a supported device.” I am using PyTorch 2.1.2, and my graphics card is 2080Ti with 22GB, which has left me quite confused. If you have time, I would appreciate your response. Once again, thank you for your excellent work.KJ您好，首先感谢您的出色工作，为我提供了训练LoRA的高效接口。但是我遇到了一个问题。当我选择fp16时，我遇到了与该问题的作者相同的情况。但当我选择 bf16 时，ComfyUI 会提示我以下消息：“bf16 混合精度需要 PyTorch >= 1.10 和支持的设备。”我使用的是 PyTorch 2.1.2，我的显卡是 2080Ti，22GB，这让我很困惑。如果您有时间，我将不胜感激您的回复。再次感谢您所做的出色工作。
> 
> Pytorch 2.1.2 is also too old and kohya himself recommends 2.4.0 minimum, I'd recommend updating to 2.4.1. Versions before that also use a lot more VRAM so it's pretty much necessary.Pytorch 2.1.2 也太旧了，kohya 本人推荐最低版本为 2.4.0，我建议更新到 2.4.1。之前的版本也使用更多的 VRAM，因此非常有必要。

Thanks for your reply. When I used PyTorch 2.4.0, I encountered a new error indicating that my device's memory was insufficient. However, after upgrading transformers to 4.44.3, it started working smoothly, and now everything is fine. 2333333333
<img width="1608" alt="8102450e9064b7d83d052e11ace1d37" src="https://github.com/user-attachments/assets/3a4e16b4-a6ee-421d-ad32-fed17d5e0fd9">
!
